src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,745 @@
+     */
+    @SuppressWarnings("unchecked")
+    private GathererOp(Gatherer<T, A, R> gatherer, GathererOp<?, ?, T> upstream) {
+        super((AbstractPipeline<?, T, ?>) upstream.upstream(),
+              upstream,
+              opFlagsFor(gatherer.integrator()));
+        this.gatherer = gatherer;
+    }
+
+    /* This allows internal access to the previous stage,
+     * to be able to fuse `gather` followed by `collect`.
+     */
+    @SuppressWarnings("unchecked")
+    private AbstractPipeline<?, T, ?> upstream() {
+        return (AbstractPipeline<?, T, ?>) super.previousStage;
+    }
+
+    @Override
+    boolean opIsStateful() {
+        // TODO
+        /* Currently GathererOp is always stateful,
+         * but it would be possible to return `false` if:
+         * - the Gatherer's initializer is Gatherer.defaultInitiatizer(),
+         * - the Gatherer's combiner is NOT Gatherer.defaultCombiner()
+         * - the Gatherer's finisher is Gatherer.defaultFinisher()
+         */
+        return true;
+    }
+
+    @Override
+    Sink<T> opWrapSink(int flags, Sink<R> downstream) {
+        return new GatherSink<>(gatherer, downstream);
+    }
+
+    /*
+     * This is used when evaluating .gather() operations interspersed with
+     * other Stream operations (in parallel)
+     */
+    @Override
+    <I> Node<R> opEvaluateParallel(PipelineHelper<R> unused1,
+                                   Spliterator<I> spliterator,
+                                   IntFunction<R[]> unused2) {
+        return this.<NodeBuilder<R>, Node<R>>evaluate(
+                upstream().wrapSpliterator(spliterator),
+                true,
+                gatherer,
+                NodeBuilder::new,
+                NodeBuilder::accept,
+                NodeBuilder::join,
+                NodeBuilder::build
+        );
+    }
+
+    @Override
+    <P_IN> Spliterator<R> opEvaluateParallelLazy(PipelineHelper<R> helper,
+                                                 Spliterator<P_IN> spliterator) {
+        /*
+         * There's a very small subset of possible Gatherers which would be
+         * expressible as Spliterators directly,
+         * - the Gatherer's initializer is Gatherer.defaultInitiatizer(),
+         * - the Gatherer's combiner is NOT Gatherer.defaultCombiner()
+         * - the Gatherer's finisher is Gatherer.defaultFinisher()
+         */
+        return opEvaluateParallel(null, spliterator, null).spliterator();
+    }
+
+    /* gather-operations immediately followed by (terminal) collect-operations
+     * are fused together to avoid having to first run the gathering to
+     * completion and only after that be able to run the collection on top of
+     * the output.  This is highly beneficial in the parallel case as stateful
+     * operations cannot be pipelined in the ReferencePipeline implementation.
+     * Overriding collect-operations overcomes this limitation.
```

Comment:
Does this mean that .parallel().gather(myGatherer).map(smth).collect(..) will be slower than .parallel().gather(myGatherer.andThen(mappingGatherer(smth))).collect(..)? Also, what about other terminals (e.g. reduce())? Will they require processing all the upstream before reduction?

Reply:
That will completely depend on many different factors, including the implementation of Stream (parallel is advisory only), the nature of the source Spliterator, the implementation of `map` and the implementation of `collect` (besides environmental factors such as availability of memory, CPU, etc).

Currently, in this PR, GathererOP is a "stateful" operation (in the reference implementation parlance), although consecutive `gather()`- operations are fused, and a `collect` after a `gather` is fused.

There is a combination of characteristics of Gatherers where they could be encoded as Spliterators, and in that case it could be conceived that they could be considered stateless. This is something which can be explored during the Preview.

There are multiple microbenchmarks in this PR for those who are curious about current performance.